tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,272 @@
+
+        change_basis = eigvecs_b @ np.diag(1 / np.sqrt(eigvals_b))
+        new_a_matrix = change_basis.T @ a_matrix @ change_basis
+        eigvals, eigvecs_a = scipy.linalg.eig(new_a_matrix)
+        eigvecs_a = self._normalize_eigvecs(eigvecs_a)
+
+        eigvecs = change_basis @ eigvecs_a
+
+        return eigvals, eigvecs
+
+    def _find_amplitudes(
+        self,
+        signal: np.ndarray[complex],
+        prev_eigvals: np.ndarray[complex],
+        eigvals: np.ndarray[complex],
+        eigvecs: np.ndarray[complex],
+    ) -> np.ndarray[complex]:
+        """Compute the amplitudes"""
+        M = int(len(signal) / 2) - 2
+        J = len(eigvals)
+        zvals = prev_eigvals / np.abs(prev_eigvals)
+        amplitudes = np.zeros(J, dtype=complex)
+
+        zinvl = np.exp(1j * np.outer(np.log(zvals) * 1j, np.arange(M + 1)))
+        factor = zinvl @ signal[: M + 1]
+
+        for k in range(J):
+            amplitudes[k] = np.square(eigvecs[:, k] @ factor)
+
+        return amplitudes
+
+    def _find_errors(
+        self,
+        eigvals: np.ndarray[complex],
+        u_matrices: np.ndarray[complex],
+        eigvecs: np.ndarray[complex],
+    ) -> np.ndarray[complex]:
+        """Find the eigenvalue error"""
+        J = len(eigvals)
+        errors = np.zeros(J)
+        for k in range(J):
+            errors[k] = np.linalg.norm(
+                (u_matrices[2] - (eigvals[k] ** 2) * u_matrices[0]) @ eigvecs[:, k]
+            )
+        return errors
+
+    def _iterate(
+        self, signal: np.ndarray[complex], prev_eigvals: np.ndarray[complex]
+    ) -> ResonanceData:
+        """Run a single iteration of the solver."""
+        u_matrices = self._evaluate_matrices(signal, prev_eigvals)
+
+        eigvals, eigvecs = self._solve_gen_eig_prob(u_matrices[1], u_matrices[0], self.rcond)
+
+        errors = self._find_errors(eigvals, u_matrices, eigvecs)
+
+        amplitudes = self._find_amplitudes(signal, prev_eigvals, eigvals, eigvecs)
+
+        return ResonanceData(eigvals=eigvals, errors=errors, complex_amplitudes=amplitudes)
+
+    def run(self, signal: np.ndarray[complex], time_step: float) -> xr.Dataset:
+        """Finds resonances in a time series.
+        Note that the signal should start after the sources have turned off.
+
+        Parameters
+        ----------
+        signal : np.ndarray[complex]
+            Time series to search for resonances
+        time_step : float
+            Time step / sampling rate of the data (in seconds).
+
+        Returns
+        -------
+        :class:`.xr.Dataset`
+            Dataset containing the frequency, decay rate, Q, amplitude, phase,
+            and estimation error of the resonances. Modes with low Q, small
+            amplitude, or high estimation error are likely to be spurious.
+        """
+        J = self.freq_grid
+        if self.freq_grid is None:
+            J = int(time_step * len(signal) * (self.freq_window[1] - self.freq_window[0]) / 2)
```

Comment:
I would suggest first unpacking
```fmin, fmax = self.freq_window```
and then using it here and lines 249 / 250, just to clean things up a little bit.